### DISCUSSION TITLE
`NEON` or `SSE*` ?

### DISCUSSION THREAD

**Discussion by ryuujo1573** [emscripten-core/emscripten#20748] [Q&A] [answered]
There's an existing code base with both arm_neon & x86 sse4.1 support, specified by conditional compiling.

As for web distribution, need I compile the codes into different binaries like `foo_x86.wasm` / `foo_arm.wasm` and distribute each according to `sec-ch-ua-arch` hints?

Or is it better to port existing code using Wasm intrinsics so as to leverage performance?

Or should I simply activate relaxed simd

**Comment by sbc100** [comment]
If there is no way to build an x86 or arm specific wasm file.   The neon and x86 sse support we have in emscripten basically attempts to emulate those standards on top of wasm simd intrinsics.    

Depending on what type of operations you are doing I imagine it could be possible for either the x86 emulation or the arm emulation to be faster, but for best performance you would likely want to use specific wasm simd.

**Reply by ryuujo1573** [reply]
If I'm getting it right, it is better to write SIMD code using Wasm intrinsics directly, and thus, there's no need to deliver arm/x86 separately and one of them would be faster in reality?

**Reply by sbc100** [reply, marked as answer]
There would never be any reason to deliver separate x86 and arm binaries, no.   In fact it would (almost) impossible for you to detect which binary to download even if you wanted to do that.

When building the with x86 or arm intrinsics the they both get lowers to Wasm intrinsics so there is no longer any bias to towards one or the other in the final wasm file.
